### ISSUE TITLE
smooth_l1 discrepancy between this implementation and original paper

### ISSUE THREAD

**Issue by daavoo** [fizyr/keras-retinanet#652]
Hi! 

I have taken a look at the losses implementation and I don't fully understand where the smooth_l1's `sigma` parameter comes from or why the default value is 0.3

I understand what the sigma parameter does but I didn't find mentions to `sigma` in the original paper formulation (RetinaNet paper says `The training loss is the sum the focal loss and the standard smooth L1 loss used for box regression [Fast-RCNN paper]`)

I also didn't find it some other smooth_1 implementations ([TensorFlow Object Detection API](https://github.com/tensorflow/models/blob/master/research/object_detection/core/losses.py#L119), [Keras Mask-RCNN](https://github.com/matterport/Mask_RCNN/blob/master/mrcnn/model.py#L1073), [Pytorch (used by pytorch-retinanet)](https://pytorch.org/docs/stable/nn.html#torch.nn.SmoothL1Loss))

However I did find mentions to sigma in [Detectron](https://github.com/caffe2/caffe2/blob/master/modules/detectron/select_smooth_l1_loss_op.cc#L36)(I guess that is what they call beta) and [mxnet](http://mxnet.apache.org/api/python/symbol/symbol.html#mxnet.symbol.smooth_l1) although in Detectron (which is supposed to be the "official" implementation) the [value used is (I think) 0.11](https://github.com/facebookresearch/Detectron/blob/2362ea3b22af8264b4e9ffdef90c161bfab8246e/detectron/core/config.py#L515).

Could anyone please clarify where the default `0.3` value comes from or what lead to include this sigma parameter?

**Comment by daavoo** [comment]
Um. I just find https://github.com/rbgirshick/py-faster-rcnn/issues/89

**Comment by daavoo** [comment]
So according to the linked issue, the author says that:

> As sigma -> inf the loss approaches L1 (abs) loss. Setting sigma = 3, makes the transition point from quadratic to linear happen at |x| <= 1 / 3**2 (closer to the origin). **The reason for doing this is because the RPN bbox regression targets are not normalized by their stdev** (unlike in Fast R-CNN)

However in retinanet (at least in this implementation) the [regression targets are normalized?](https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/utils/anchors.py#L315)

**Comment by hgaiser** [comment]
The default for sigma is `3.0`, not `0.3`, but I assume that is a typo? The `3.0` is actually the same as the `0.11` used in Detectron. The difference comes from what it means. In Detectron they use it directly as a threshold, in keras-retinanet we based it on the implementation of py-faster-rcnn, which uses `1 / sigma**2`, which is `1 / 9 = 0.1111...`. So really both values are more or less identical.

Regarding the comment of https://github.com/rbgirshick/py-faster-rcnn/issues/89, I'm not a 100% sure on this but I believe in FastRCNN the regression targets were normalized by their *current* stddev and in FasterRCNN they used the *static* stddev precomputed from COCO (which is also what we do). I don't know exactly where it happens, I think it was in the bbox_inside_weights or bbox_outside_weights or something, but py-faster-rcnn also does the normalization by mean and stddev.

Let me know if any of this makes sense :p

**Comment by daavoo** [comment]
Thanks for the clarifications. It does makes sense. I don't know if the sigma parameter has a big impact on the convergence but it looks like a thing that should have been mentioned on the paper
